storage/tsdb/tsdb.go

```diff
@@ -190,14 +190,39 @@ type querier struct {
 
 func (q querier) Select(oms ...*labels.Matcher) (storage.SeriesSet, error) {
 	ms := make([]tsdbLabels.Matcher, 0, len(oms))
+	var err error
+	notSeen := false
 
-	for _, om := range oms {
+	for i, om := range oms {
 		ms = append(ms, convertMatcher(om))
+
+		// See: https://github.com/prometheus/prometheus/issues/3575
+		if om.Type == labels.MatchNotRegexp || om.Type == labels.MatchNotEqual {
+			if notSeen == false {
```

Comment:
This doesn't seem right, there can be multiple negative matchers.

Reply:
This is right because there is a recursive call that handles other matchers.